File: backend/app/api/cityseo.py
"""City pillar pages for the Local SEO universe, served at
/{locale}/local-seo-services/{city-slug} (for example /en-in/local-seo-services/mumbai).

Storage deliberately reuses `lpseo_pages`. A city pillar is a leaf in the same URL
namespace as the industry x city pages, so giving it its own table would mean two
tables racing for one slug and a second resolver on the hot public route. Instead the
row lands in lpseo_pages with slug == industry_slug == city_slug (see
PLACEHOLDER_INDUSTRY_LABEL for why the industry mirrors the slug), and the EXISTING
/public/lpseo/{slug} endpoint serves it unchanged. `content["page_type"]`
is the discriminator the frontend branches on. No new table, no migration.

Why this module exists at all, rather than more columns in lpseo.py: the city-pillar
package ships a genuinely different content schema. Its CSV packs JSON arrays and
JSON objects inside single cells instead of the "|" / "::" separator scheme every
other package uses, and its sections (catchments, operating-model table, 90 day plan,
partner questions) have no counterpart on an industry x city leaf. Duplicating the
importer keeps that difference explicit instead of overloading a shared parser.

Deliberate omissions:
- `schema_json` is IGNORED. Schema is generated server-side from the visible copy, so
  the FAQPage block cannot drift from the FAQs a visitor actually sees, and a package
  that ships stale or invalid JSON-LD cannot publish it straight to production.
- `form_destination` is IGNORED. The package points at formsubmit.co, which neither
  stores the lead nor survives QA. The rendered page reuses the shared lead form,
  which posts to /public/lpseo/leads and persists before emailing.
- `robots_status` is READ BUT NOT OBEYED. Every import lands draft + noindex; going
  live is a separate, deliberate admin action once launch QA passes.
"""
import json
import logging
import re
from typing import Any, Optional

from fastapi import APIRouter, Depends, HTTPException, UploadFile, File
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.api.deps import superadmin_required
from app.core.redis_client import get_redis
from app.models.lpseo_page import LpseoPage, LpseoPageStatus
from app.models.user import User
from app.services.revalidation_service import trigger_lpseo_page_flush
# Page-type-agnostic file reader, reused verbatim from pseo.
from app.api.pseo import _read_rows

logger = logging.getLogger(__name__)

admin_router = APIRouter()

SITE_URL = "https://www.pinzo.io"

# Shared vocabulary with lpseo_pages.page_type ("leaf" / "industry_pillar" /
# "city_pillar"). Written in two places on purpose:
#   - the COLUMN, so the drip scheduler can say "leaves only" and never trickle a
#     hand-authored pillar into the index;
#   - content["page_type"], because the public payload serialises `content` but not
#     the column, and the frontend needs the discriminator to pick a template.
PAGE_TYPE = "city_pillar"

# A city pillar has no industry, but lpseo_pages.industry_slug is NOT NULL and indexed
# with city_slug, so it needs a value.
#
# That value is the CITY SLUG, i.e. the row's own slug, and that choice is load-bearing.
# /{locale}/local-seo-services/{seg} resolves a page by slug first and only falls back
# to "industry hub listing every published page with industry_slug == seg" when no page
# owns the slug. A shared sentinel like "city-pillar" is therefore a live URL:
# /en-in/local-seo-services/city-pillar owns no page, so a published + indexed pillar
# would make it render an indexable phantom hub titled "City Pillar", and the sitemap
# (which emits one hub URL per distinct industry_slug) would advertise it. Setting
# industry_slug == slug makes the only URL that can reach the hub branch the pillar's
# own URL, which always resolves to the pillar itself first. It also makes the shared
# isLpseoPillar() check (slug === industry_slug) true, which is what keeps the pillar
# out of the country hub's industry cards and out of the sitemap's leaf list.
#
# The label stays a constant: it is admin-facing only. With industry_slug == slug the
# frontend never renders it, because every path that would have is now unreachable.
PLACEHOLDER_INDUSTRY_LABEL = "City Pillar"

# lpseo_pages.page_type is arriving alongside this module. Set it when the model has
# it, so neither change has to land before the other to keep imports working.
_HAS_PAGE_TYPE_COLUMN = hasattr(LpseoPage, "page_type")


# ── Column schema ─────────────────────────────────────────────────────────────
#
# This package packs JSON inside cells. A list column holds a JSON array of strings,
# an object-list column a JSON array of objects, an object column a JSON object. That
# is the package's own convention and the enriched CSV keeps it.

IDENTITY_COLS = [
    "locale", "canonical_url", "meta_title", "meta_description", "h1",
    "city_label", "city_slug", "country", "quality_score", "robots_status",
]

TEXT_COLS = [
    "page_id", "page_type", "primary_keyword", "hero_eyebrow", "hero_copy",
    "hero_image_alt", "hero_image_caption", "primary_cta", "secondary_cta",
    "direct_question", "direct_answer", "direct_answer_support",
    "city_context", "catchment_heading", "catchment_intro",
    "service_scope_heading", "ranking_heading",
    "operating_model_heading", "roadmap_heading", "roadmap_note",
    "proof_heading", "proof_intro", "proof_image_alt", "proof_asset_url",
    "ai_heading", "ai_signals_heading", "ai_entity_plan",
    "industry_links_heading", "industry_links_intro", "partner_heading",
    "faq_heading", "lead_eyebrow", "lead_heading", "lead_intro", "lead_consent_note",
    "og_title", "og_description", "og_image", "og_image_alt", "hreflang_x_default",
    "country_label", "state_region", "parent_url",
]

# JSON array of strings.
LIST_COLS = [
    "secondary_keywords", "industry_search_journeys", "industry_entities",
    "service_matrix", "proof_asset_type", "hero_trust_points", "ranking_intro",
    "ai_intro", "ai_entity_signals", "operating_model_columns", "lead_points",
]

# JSON array of objects. The value is the expected key set: a row missing a key is a
# broken import, not a half-filled section, because one rule for this page type is
# that a section is filled entirely by imported content or not rendered at all.
OBJECT_LIST_COLS: dict[str, tuple[str, ...]] = {
    "internal_links": ("anchor", "url"),
    "catchment_areas": ("name", "detail"),
    "service_scope": ("title", "detail"),
    "ranking_factors": ("title", "detail"),
    "operating_model_rows": ("area", "single", "multi"),
    "roadmap_phases": ("days", "title", "steps"),
    "proof_metrics": ("label", "detail"),
    "industry_link_groups": ("title", "links"),
    "partner_questions": ("title", "detail"),
    "faqs": ("q", "a"),
}

# JSON object.
OBJECT_COLS = ["section_eyebrows"]

# Read for context but never applied. See the module docstring.
IGNORED_COLS = ["schema_json", "form_destination"]

ALL_COLS = (IDENTITY_COLS + TEXT_COLS + LIST_COLS
            + list(OBJECT_LIST_COLS) + OBJECT_COLS + IGNORED_COLS)


# ── Cell parsing ──────────────────────────────────────────────────────────────

def _loads(col: str, cell: str) -> Any:
    try:
        return json.loads(cell)
    except json.JSONDecodeError as e:
        raise ValueError(f"{col}: cell is not valid JSON ({e.msg} at char {e.pos})")


def _json_list(col: str, cell: str) -> list[str]:
    val = _loads(col, cell)
    if not isinstance(val, list):
        raise ValueError(f"{col}: expected a JSON array, got {type(val).__name__}")
    return [str(v).strip() for v in val if str(v).strip()]


def _json_object_list(col: str, cell: str, keys: tuple[str, ...]) -> list[dict]:
    val = _loads(col, cell)
    if not isinstance(val, list):
        raise ValueError(f"{col}: expected a JSON array, got {type(val).__name__}")
    out = []
    for i, item in enumerate(val):
        if not isinstance(item, dict):
            raise ValueError(f"{col}[{i}]: expected an object, got {type(item).__name__}")
        missing = [k for k in keys if k not in item]
        if missing:
            raise ValueError(f"{col}[{i}]: missing {', '.join(missing)}")
        out.append({k: item[k] for k in keys})
    return out


def _json_object(col: str, cell: str) -> dict:
    val = _loads(col, cell)
    if not isinstance(val, dict):
        raise ValueError(f"{col}: expected a JSON object, got {type(val).__name__}")
    return val


# ── Identity derivation ───────────────────────────────────────────────────────
#
# The package CSV carries no city_label, city_slug or country column. Rather than
# make an operator hand-add three columns to every city package, derive them and let
# an explicit column win when one is present.

_H1_CITY_RE = re.compile(r"^\s*Local SEO Services in\s+(.+?)\s*$", re.IGNORECASE)


def _slugify(value: str) -> str:
    return re.sub(r"-{2,}", "-", re.sub(r"[^a-z0-9]+", "-", value.strip().lower())).strip("-")


def _derive_city_label(row: dict[str, str]) -> str:
    explicit = (row.get("city_label") or "").strip()
    if explicit:
        return explicit
    m = _H1_CITY_RE.match(row.get("h1") or "")
    if m:
        return m.group(1)
    raise ValueError("city_label is required (and could not be derived from h1)")


def _derive_city_slug(row: dict[str, str], city_label: str) -> str:
    explicit = (row.get("city_slug") or "").strip().lower()
    if explicit:
        return explicit
    # The canonical URL is the authority on the published slug; the label is only a
    # fallback, because a label and a slug can legitimately differ ("Bengaluru").
    canonical = (row.get("canonical_url") or "").strip().rstrip("/")
    if canonical:
        last = canonical.rsplit("/", 1)[-1]
        if last and last != "local-seo-services":
            return last.lower()
    return _slugify(city_label)


def _derive_country(row: dict[str, str]) -> str:
    explicit = (row.get("country") or "").strip().lower()
    if len(explicit) == 2:
        return explicit
    locale = (row.get("locale") or "").strip().lower()
    if re.fullmatch(r"[a-z]{2}-[a-z]{2}", locale):
        return locale.split("-")[1]
    return "in"


# ── Row -> page ───────────────────────────────────────────────────────────────

class CityPageIn(BaseModel):
    slug: str
    country: str
    city_label: str
    city_slug: str
    meta_title: str
    meta_description: str
    h1: str
    canonical_url: Optional[str] = None
    quality_score: Optional[int] = None
    content: dict[str, Any] = Field(default_factory=dict)


def parse_city_row(row: dict[str, str]) -> CityPageIn:
    """Turn one enriched-CSV row into the page we will store. Pure: no DB, no IO."""
    row = {(k or "").strip().lower(): (v or "") for k, v in row.items()}

    content: dict[str, Any] = {}
    for col in TEXT_COLS:
        if row.get(col, "").strip():
            content[col] = row[col].strip()
    for col in LIST_COLS:
        if row.get(col, "").strip():
            content[col] = _json_list(col, row[col])
    for col, keys in OBJECT_LIST_COLS.items():
        if row.get(col, "").strip():
            content[col] = _json_object_list(col, row[col], keys)
    for col in OBJECT_COLS:
        if row.get(col, "").strip():
            content[col] = _json_object(col, row[col])

    # The discriminator the public route branches on. Stamped unconditionally so a
    # row whose page_type cell was blank or misspelled still renders as a city pillar
    # rather than silently falling through to the industry template.
    content["page_type"] = PAGE_TYPE

    city_label = _derive_city_label(row)
    city_slug = _derive_city_slug(row, city_label)
    country = _derive_country(row)
    locale = f"en-{country}"

    qs = row.get("quality_score", "").strip()
    try:
        quality_score = int(float(qs)) if qs else None
    except ValueError:
        quality_score = None

    h1 = row.get("h1", "").strip() or f"Local SEO Services in {city_label}"
    return CityPageIn(
        slug=city_slug,
        country=country,
        city_label=city_label,
        city_slug=city_slug,
        meta_title=row.get("meta_title", "").strip() or f"Local SEO Services in {city_label} | Pinzo",
        meta_description=row.get("meta_description", "").strip() or (
            f"Local SEO services in {city_label} for stronger Google Maps visibility, "
            f"local rankings, reviews and leads."
        ),
        h1=h1,
        canonical_url=row.get("canonical_url", "").strip() or f"{SITE_URL}/{locale}/local-seo-services/{city_slug}/",
        quality_score=quality_score,
        content=content,
    )


# ── Cache invalidation ────────────────────────────────────────────────────────
#
# The row is served by /public/lpseo/{slug}, so it sits under that endpoint's Redis
# key. Keep this string in sync with lpseo._cache_key; it is duplicated rather than
# imported so a refactor there cannot silently reach into this module's behaviour.
def _invalidate(slug: str, country: str) -> None:
    try:
        get_redis().delete(f"public_lpseo:{slug}")
    except Exception:
        pass
    # Narrow flush: a city pillar owns one path and busts no list tag.
    trigger_lpseo_page_flush(slug, country)


# ── Admin ─────────────────────────────────────────────────────────────────────

@admin_router.get("/import/columns")
def admin_import_columns(_: User = Depends(superadmin_required)):
    return {
        "columns": ALL_COLS,
        "required": ["h1"],
        "cell_format": "JSON inside cells: list columns hold a JSON array of strings, "
                       "object-list columns a JSON array of objects, object columns a JSON object.",
        "identity_columns": IDENTITY_COLS,
        "text_columns": TEXT_COLS,
        "list_columns": LIST_COLS,
        "object_list_columns": {k: list(v) for k, v in OBJECT_LIST_COLS.items()},
        "object_columns": OBJECT_COLS,
        "ignored_columns": IGNORED_COLS,
        "derived_columns": ["city_label (from h1)", "city_slug (from canonical_url)", "country (from locale)"],
        "notes": [
            "schema_json is ignored; JSON-LD is generated from the visible copy.",
            "form_destination is ignored; the page uses the stored /public/lpseo/leads form.",
            "robots_status is ignored; every import lands draft + noindex until launch QA passes.",
        ],
    }


@admin_router.post("/import")
def admin_import(file: UploadFile = File(...), db: Session = Depends(get_db),
                 _: User = Depends(superadmin_required)):
    """Import city pillars from the package CSV/XLSX into lpseo_pages.

    Always draft + noindex, whatever robots_status says. An imported pillar is
    unreviewed content: publishing and indexing are separate admin actions taken
    after the launch checklist passes.
    """
    blob = file.file.read()
    if len(blob) > 10 * 1024 * 1024:
        raise HTTPException(status_code=400, detail="File too large (max 10 MB)")
    rows = _read_rows(file.filename or "import.csv", blob)

    results, created, updated, failed, touched = [], 0, 0, 0, []
    for idx, row in enumerate(rows, start=2):
        if not any((v or "").strip() for v in row.values()):
            continue
        try:
            data = parse_city_row(row)
            page = db.query(LpseoPage).filter(LpseoPage.slug == data.slug).first()
            if page is not None:
                # Refuse to overwrite a real industry x city leaf that happens to own
                # this slug. Silently converting /dentists-in-mumbai into a city
                # pillar would take a live page down.
                existing_type = (getattr(page, "page_type", None)
                                 or (page.content or {}).get("page_type") or "")
                if str(existing_type).strip().lower() != PAGE_TYPE:
                    raise ValueError(
                        f"slug '{data.slug}' already belongs to a non-city-pillar page "
                        f"({page.industry_slug or 'unknown'}); refusing to overwrite"
                    )
                was_published = page.status == LpseoPageStatus.PUBLISHED.value
                # Rewritten on every update, not only on create: a row imported before
                # industry_slug became the city slug still carries the old sentinel,
                # and leaving it there leaves the phantom hub URL live.
                page.industry_label = PLACEHOLDER_INDUSTRY_LABEL
                page.industry_slug = data.city_slug
                page.city_label, page.city_slug, page.country = data.city_label, data.city_slug, data.country
                page.meta_title, page.meta_description, page.h1 = data.meta_title, data.meta_description, data.h1
                page.canonical_url = data.canonical_url or None
                page.quality_score, page.content = data.quality_score, data.content
                page.index_status = "noindex"
                page.status = LpseoPageStatus.DRAFT.value
                if _HAS_PAGE_TYPE_COLUMN:
                    page.page_type = PAGE_TYPE
                action = "updated"
                updated += 1
            else:
                was_published = False
                page = LpseoPage(
                    slug=data.slug,
                    industry_label=PLACEHOLDER_INDUSTRY_LABEL,
                    industry_slug=data.city_slug,
                    city_label=data.city_label, city_slug=data.city_slug, country=data.country,
                    meta_title=data.meta_title, meta_description=data.meta_description, h1=data.h1,
                    canonical_url=data.canonical_url or None,
                    index_status="noindex", quality_score=data.quality_score,
                    content=data.content, status=LpseoPageStatus.DRAFT.value,
                    **({"page_type": PAGE_TYPE} if _HAS_PAGE_TYPE_COLUMN else {}),
                )
                db.add(page)
                action = "created"
                created += 1
            db.flush()
            if was_published:
                touched.append((page.slug, page.country))
            results.append({"row": idx, "slug": page.slug, "action": action,
                            "status": page.status, "index_status": page.index_status})
        except Exception as e:
            db.rollback()
            failed += 1
            results.append({"row": idx, "slug": (row.get("city_slug") or "").strip() or None,
                            "action": "error", "error": str(e)})
    db.commit()
    for slug, country in set(touched):
        _invalidate(slug, country)
    return {"created": created, "updated": updated, "failed": failed, "results": results}

File: frontend/lib/cityseo.ts
// City pillar for the Local SEO universe, served at
// /{locale}/local-seo-services/{city-slug} (for example /en-in/local-seo-services/mumbai).
//
// It shares STORAGE with the industry x city leaf (both are rows in lpseo_pages, both
// come back from /public/lpseo/{slug}) but not its content schema: the city package
// ships catchments, an operating-model table, a 90 day plan and partner questions that
// have no counterpart on a leaf. So the shape lives here and the page is rendered by
// components/cityseo/CityPillar.tsx, discriminated on content.page_type.

export interface CityLink { anchor: string; url: string }
export interface CityQa { q: string; a: string }
export interface CityArea { name: string; detail: string }
export interface CityPair { title: string; detail: string }
export interface CityService { title: string; detail: string; num?: string }
export interface CityOperatingRow { area: string; single: string; multi: string }
export interface CityPhase { days: string; title: string; steps: string[] }
export interface CityMetric { label: string; detail: string }
export interface CityLinkGroup { title: string; links: CityLink[] }

export type CitySectionKey =
  | 'answer' | 'catchments' | 'scope' | 'ranking' | 'operating_model' | 'roadmap'
  | 'proof' | 'ai' | 'industry_links' | 'partner' | 'faq'

export interface CityContent {
  // Discriminator stamped by the importer. Everything else is optional because a
  // section renders only when the import filled it.
  page_type?: string

  page_id?: string
  primary_keyword?: string
  section_eyebrows?: Partial<Record<CitySectionKey, string>>

  // Hero
  hero_eyebrow?: string
  hero_copy?: string
  hero_image_alt?: string
  hero_image_caption?: string
  hero_trust_points?: string[]
  primary_cta?: string
  secondary_cta?: string

  // Direct answer
  direct_question?: string
  direct_answer?: string
  direct_answer_support?: string

  // Catchments
  catchment_heading?: string
  catchment_intro?: string
  catchment_areas?: CityArea[]

  // Service scope. `service_matrix` is the package's flat title list; `service_scope`
  // is the same nine areas WITH the descriptions recovered from the reference HTML.
  service_scope_heading?: string
  service_scope?: CityService[]
  service_matrix?: string[]

  // Relevance / distance / prominence
  ranking_heading?: string
  ranking_intro?: string[]
  ranking_factors?: CityPair[]

  // One location vs many
  operating_model_heading?: string
  operating_model_columns?: string[]
  operating_model_rows?: CityOperatingRow[]

  // 90 day plan
  roadmap_heading?: string
  roadmap_phases?: CityPhase[]
  roadmap_note?: string

  // Evidence and reporting
  proof_heading?: string
  proof_intro?: string
  proof_image_alt?: string
  proof_metrics?: CityMetric[]
  proof_asset_type?: string[]
  proof_asset_url?: string

  // AI search readiness
  ai_heading?: string
  ai_intro?: string[]
  ai_signals_heading?: string
  ai_entity_signals?: string[]
  ai_entity_plan?: string

  // Industry child pages
  industry_links_heading?: string
  industry_links_intro?: string
  industry_link_groups?: CityLinkGroup[]
  internal_links?: CityLink[]

  // Partner selection
  partner_heading?: string
  partner_questions?: CityPair[]

  // FAQs. Visible copy AND the only source of the FAQPage schema.
  faq_heading?: string
  faqs?: CityQa[]

  // Lead capture
  lead_eyebrow?: string
  lead_heading?: string
  lead_intro?: string
  lead_points?: string[]
  lead_consent_note?: string

  // Search-strategy context (kept from the package, not rendered as body copy).
  // `city_context` belongs here, not in the catchment section: it is the package's
  // one-paragraph summary of the catchment thesis, and its first sentence is
  // verbatim the catchment_heading, so rendering it echoes the H2 immediately below
  // it. The reference HTML never shows it either.
  city_context?: string
  secondary_keywords?: string[]
  industry_search_journeys?: string[]
  industry_entities?: string[]

  // Meta
  og_title?: string
  og_description?: string
  og_image?: string
  og_image_alt?: string
  hreflang_x_default?: string
  country_label?: string
  state_region?: string
  parent_url?: string
}

// Identity fields, all of which an LpseoPageData already carries.
export interface CityPageBase {
  slug: string
  country: string
  locale: string
  city_label: string
  city_slug: string
  meta_title: string
  meta_description: string
  h1: string
  canonical_url: string | null
  index_status: string
  quality_score: number | null
  published_at: string | null
  updated_at: string | null
}

export interface CityPageData extends CityPageBase {
  content: CityContent
}

// The router runs with trailingSlash:false, so every slashed URL 308-redirects.
// The content packages author canonicals WITH a slash, which made three tiers
// self-canonicalise to a redirect. The router wins: strip it at the edge so the
// canonical, og:url, hreflang and every JSON-LD @id agree with what actually serves.
export const noSlash = (u: string) => (u.length > 1 && u.endsWith('/') ? u.replace(/\/+$/, '') : u)

const SITE_URL = process.env.NEXT_PUBLIC_APP_URL || 'https://www.pinzo.io'

export const CITY_SEGMENT = 'local-seo-services'
export const cityPath = (locale: string, slug: string) => `/${locale}/${CITY_SEGMENT}/${slug}`
export const cityUrl = (locale: string, slug: string) => `${SITE_URL}${cityPath(locale, slug)}`
export const countryHubPath = (locale: string) => `/${locale}/${CITY_SEGMENT}`

// Same vocabulary as lpseo_pages.page_type ("leaf" / "industry_pillar" / "city_pillar").
// The public payload serialises `content` but not the column, so the discriminator
// the browser sees is content.page_type.
const CITY_PILLAR = 'city_pillar'

/** True when a page fetched from /public/lpseo/{slug} is a city pillar rather than an
 *  industry x city leaf. Both live in the same table; this is the discriminator. */
export function isCityPillar(page: { content?: unknown } | null | undefined): boolean {
  const t = (page?.content as { page_type?: string } | undefined)?.page_type
  return (t || '').trim().toLowerCase() === CITY_PILLAR
}

/** Narrow an lpSEO payload to the city-pillar shape. One cast, in one place, guarded
 *  by isCityPillar at the call site. */
export function asCityPage(page: CityPageBase & { content?: unknown }): CityPageData {
  return { ...page, content: (page.content || {}) as CityContent }
}

// Combined @graph: Organization, Service, WebPage, BreadcrumbList, and FAQPage when
// the page has FAQs.
//
// The FAQPage block is built FROM `content.faqs`, the exact array the component
// renders, so the visible answer and the structured answer are the same string by
// construction. The package's own `schema_json` column is never imported and never
// read here; a hand-written schema block is precisely how visible copy and markup
// drift apart.
export function buildCityJsonLd(page: CityPageData) {
  const url = noSlash(page.canonical_url || cityUrl(page.locale, page.slug))
  const orgId = `${SITE_URL}/#organization`
  const c = page.content || {}
  const region = c.state_region
  const country = c.country_label

  const graph: Record<string, unknown>[] = [
    {
      '@type': 'Organization',
      '@id': orgId,
      name: 'Pinzo',
      url: SITE_URL,
      logo: `${SITE_URL}/logo-horizontal-3.png`,
    },
    {
      // Kept in step with the rest of the family's schema baseline.
      '@type': 'WebSite',
      '@id': `${SITE_URL}/#website`,
      url: SITE_URL,
      name: 'Pinzo',
      publisher: { '@id': orgId },
    },
    {
      '@type': 'Service',
      '@id': `${url}#service`,
      name: `Local SEO Services in ${page.city_label}`,
      serviceType: 'Local search engine optimisation',
      provider: { '@id': orgId },
      // Truthful service area: the city, plus its region when the import states one.
      // No LocalBusiness and no address: Pinzo has no eligible location in this city.
      areaServed: [
        { '@type': 'City', name: page.city_label },
        ...(region ? [{ '@type': 'AdministrativeArea', name: region }] : []),
      ],
      audience: { '@type': 'BusinessAudience', audienceType: 'Local and multi location businesses' },
      description: page.meta_description,
      url,
    },
    {
      '@type': 'WebPage',
      '@id': `${url}#webpage`,
      url,
      name: page.meta_title,
      description: page.meta_description,
      inLanguage: page.locale,
      isPartOf: { '@id': `${SITE_URL}/#website` },
      about: { '@id': `${url}#service` },
      ...(page.published_at ? { datePublished: page.published_at } : {}),
      ...(page.updated_at ? { dateModified: page.updated_at } : {}),
    },
    {
      '@type': 'BreadcrumbList',
      '@id': `${url}#breadcrumb`,
      // Must mirror the visible trail exactly, node for node and URL for URL:
      // Home > Local SEO Services (global) > Country > City. The global node has NO
      // trailing slash, matching both the visible <Link href="/local-seo-services">
      // and the same node in cseo.ts / lpseo.ts; a trailing slash here would be a
      // different URL from the one the visitor can click.
      itemListElement: [
        { '@type': 'ListItem', position: 1, name: 'Home', item: SITE_URL },
        { '@type': 'ListItem', position: 2, name: 'Local SEO Services', item: `${SITE_URL}/${CITY_SEGMENT}` },
        ...(country
          ? [{ '@type': 'ListItem', position: 3, name: country, item: `${SITE_URL}${countryHubPath(page.locale)}` }]
          : []),
        { '@type': 'ListItem', position: country ? 4 : 3, name: page.city_label, item: url },
      ],
    },
  ]

  const faqs = c.faqs || []
  if (faqs.length > 0) {
    graph.push({
      '@type': 'FAQPage',
      '@id': `${url}#faq`,
      mainEntity: faqs.map((f) => ({
        '@type': 'Question',
        name: f.q,
        acceptedAnswer: { '@type': 'Answer', text: f.a },
      })),
    })
  }
  return { '@context': 'https://schema.org', '@graph': graph }
}
